DOC: Use nbsphinx

TomAugspurger opened the pull request:
Update header levels for nbsphinx

Link to nb, nicer default table

Closes #15539

TomAugspurger committed:
DOC: Use nbsphinx for notebook doc build

Adds a new doc-dependency nbsphinx for converting jupyter notebooks
to ReST, which works better with the sphinx conversion process.
Remvoes the hacky notebook -> HTML -> raw include we had before.

TomAugspurger commented:
Here's a preview

![screen shot 2017-03-05 at 12 28 50 pm](https://cloud.githubusercontent.com/assets/1312546/23590087/573c8260-019f-11e7-982e-12caa84ae238.png)

The doc build may fail to convert the notebook if we don't have pandoc on the CI server. It's available through conda-forge if we need, but wanted to test without first.

jorisvandenbossche commented:
Why does it need pandoc?

You also included two versions of the notebook, was that on purpose? 
I also don't see where you included it in any toctree.

But seems to look much nicer!

TomAugspurger commented:
> Why does it need pandoc?

Oh, I think it's the same as nbconvert, so nothing new. 

> You also included two versions of the notebook, was that on purpose?

Nope, fixed.

> I also don't see where you included it in any toctree.

We had the dummy `style.rst` file that just included the converted html in a raw block. I removed `style.rst` and renamed the notebook to `style.ipynb`, so it gets picked up by the `style` already in the `index.rst` toctree.

jreback commented:
@TomAugspurger you need to add the install of ``nbsphinx`` in ``ci/requirements-3.5_DOC_BUILD.sh`` (you can add as same line as feather-format). That's the only place we explicitly install from conda-forge.

jreback commented on ci/install_travis.sh:
```diff
@@ -114,6 +114,10 @@ if [ "$COVERAGE" ]; then
     pip install coverage pytest-cov
 fi
 
+if [ "$DOC_BUILD" ]; then
```
move instead to ``ci/requirements-3.5_DOC_BUILD.sh`` (and I would install install via conda-forge channel (just like feather-format)

jreback commented on doc/README.rst:
```diff
@@ -81,7 +81,9 @@ have ``sphinx`` and ``ipython`` installed. `numpydoc
 this because a local copy of ``numpydoc`` is included in the pandas source
-code.
+code. `nbsphinx <https://nbsphinx.readthedocs.io/>`_ is used to convert
+Jupyter notebooks. You will need to install it if you intend to modify any of
```
should prob add this to the requirements_docs (which should also have conda-forge as a channel in the instructions)

TomAugspurger commented:
Fixed, thanks. Also added seaborn to the doc build (used in the notebook). I set the option to allow errors when executing the notebook, as that seems more consistent with the rest of the doc build.

jorisvandenbossche commented:
@TomAugspurger doc build is failing on travis due to missing pandoc. So apparently this was previously not needed.

jorisvandenbossche commented on doc/source/themes/nature_with_gtoc/static/nature.css_t:
```diff
@@ -321,3 +321,27 @@ div.seealso dd {
     margin-bottom: 0;
 }
+
+table {
```
There is some styling for rendered_html tables above. Should that then be removed? (I think that was added for the converted styling notebook)

Also, would this also influence the tables in the other docs? So maybe this should have a 'dataframe' selector?

jreback commented on ci/requirements_all.txt:
```diff
@@ -3,6 +3,7 @@ pytest-cov
 pytest-xdist
 flake8
 sphinx
```
hmm, is this our catch all for what people should install to get a dev env? should add seaborn then

TomAugspurger replied:
We also have `requirements_dev.txt`, which I think is more appropriate. I read `requirements_all.txt` as all the optional deps for using pandas (xlrd, xlwt, tables, etc.). Maybe `sphinx` and the test libs should go in `requirements_dev.txt`?

jreback replied:
yeah I think these should be more clear what the purpose of these are. You can add a comment (or 2) at the top of the file I think (# is ignored)

TomAugspurger commented:
> @TomAugspurger doc build is failing on travis due to missing pandoc. So apparently this was previously not needed.

I'd rather not have pandoc as a required dependency to build the docs. I'll see what I can do to make this fail gracefully without pandoc installed.

jorisvandenbossche commented:
Previously, we were using nbconvert to just convert it to html, which did not needed pandoc. But with nbsphinx I suppose this converts it to rst under the hood and therefore uses pandoc? (not sure if that is a correct assessment, the docs of nbsphinx are not very clear on the pandoc requirement in their installation notes) 
But, if that is correct, and if we want to include this in our dev doc build, then we just need to install pandoc? (you can add it to the conda-forge packages to install)

But indeed, it would be nice that this would not fail the doc build if you do that locally without having pandoc installed.

jreback commented:
this lgtm. @jorisvandenbossche had some comments though.

TomAugspurger commented:
Oh, whoops, forgot I had already made a merge request. I'll let you know when this is ready for review.

Quick update though: I raised an issue on nbsphinx about gracefully failing if there's no pandoc. I think they're open, but 1.) they have a stalled PR removing the need for pandoc entirely, and 2.) it's a bit tricky since the errors are raised on each markdown cell (do you skip the cell, or the entire nb?)

So I think it's probably best we just workaround it for now by removing the ipynb file(s) before calling out to sphinx. Also, should we do something like my `maybe_exclude_notebooks` for the regular `.rst` files? It can be a bit strange for a new contributor to have a bunch of files removed when you tell them to use `--single=`.

TomAugspurger commented:
Are people OK with using `requirements_all.txt` to contain all the optional runtime deps to use pandas, and `requirements_dev.txt` to contain all the optional deps to build docs and run tests?

Pip let's you source additional requirements files from another like so

```
# ci/requirements_dev.txt
# All additional dependencies required for running tests, building docs
# Main requirements are in ci/requirements_all.txt
-r requirements_all.txt
# Building
cython
# Testing
flake8
pytest
pytest-cov
pytest-xdist
# Doc building
sphinx
nbsphinx
nbconvert
ipython
seaborn
```

Does anyone know if this is possible with conda? If not I'll open an issue requesting it.

TomAugspurger commented:
Reading a bit more, it seems like `requirements_dev.txt` was intended to be a minimal set of dependencies to develop pandas. I think it's a bit strange the doc dependencies (at least sphinx) aren't in there, but the test dependencies are.

One additional option is

- `ci/requirements_base.txt`: Required deps only (plus cython?)
- `ci/requirements_dev.txt`: Test and doc deps (and cython if it's not in base)
- `ci/requirements_all.txt`: all optional deps

And then the conda install instructions are

```bash
conda env create -n pandas_dev python=3 -f ci/requirements_base.txt
conda install -n pandas_dev --file=ci/requirements_all.txt
conda install -n pandas_dev --file=ci/requirements_dev.txt
```

jreback commented:
> ci/requirements_dev.txt: Test and doc deps (and cython if it's not in base)

should also have a requirements_docs.txt I think (with only the doc-stuff).

TomAugspurger commented:
Looks like conda is [close](https://github.com/conda/conda/pull/3969) to allowing `-r` syntax in a `requirements.txt`. I'm going to just add nbsphinx to `requirements_all.txt` for now, and open another issue for reorganizing the `requirements_*.txt` files.

mgeier commented on ci/requirements-3.5_DOC_BUILD.sh:
```diff
@@ -6,6 +6,6 @@ echo "[install DOC_BUILD deps]"
 pip install pandas-gbq
 
-conda install -n pandas -c conda-forge feather-format
+conda install -n pandas -c conda-forge feather-format nbsphinx pypandoc
```
What do you need `pypandoc` for?
Wouldn't `pandoc` be more appropriate instead?
https://anaconda.org/conda-forge/pandoc

TomAugspurger replied:
Whoops, your correct. I forgot they were packaged separately. Will update.

mgeier commented on doc/make.py:
```diff
@@ -106,106 +106,42 @@ def clean():
+    yield
+    for nb, content in contents.items():
+        with open(nb, 'wt') as f:
+            f.write(content)
```
Am I seeing correctly that you are writing the file(s) each time after Sphinx is run?
This will force re-parsing each time you run it again ...
Otherwise, Sphinx would simply skip over the already built (and up-to-date) notebooks.

Also, if somebody is editing the notebook(s) at the same time, this might cause problems, right?

TomAugspurger replied:
> Am I seeing correctly that you are writing the file(s) each time after Sphinx is run?

If nbsphinx is installed then `contents` will be empty. So we should only be writing stuff if `nbsphinx` is not installed, and sphinx isn't parsing these files anyway.

For editing the notebooks at the same time, I don't think there are any additional problems with this approach.

mgeier replied:
OK, I see now. It still seems a bit brutal to me and I'm generally skeptical about writing to the source directory, but I guess it's OK.

mgeier commented:
@jorisvandenbossche 

> But with nbsphinx I suppose this converts it to rst under the hood and therefore uses pandoc? (not sure if that is a correct assessment, the docs of nbsphinx are not very clear on the pandoc requirement in their installation notes).

Yes, this is correct. At least until https://github.com/spatialaudio/nbsphinx/issues/36 is done (which is a biggie).

You are also right that the documentation doesn't mention that, but we had at least an issue for that: https://github.com/spatialaudio/nbsphinx/issues/72.
I just created a PR which hopefully closes this issue: https://github.com/spatialaudio/nbsphinx/pull/102.

Looking forward to Jupyter notebooks in the `pandas` docs!

jorisvandenbossche commented on doc/source/themes/nature_with_gtoc/static/nature.css_t:
```diff
@@ -299,25 +299,33 @@ td.field-body blockquote {
-
-div.seealso dd {
-    margin-top: 0;
-    margin-bottom: 0;
```
Should this have been removed? (not sure if it does something useful)

TomAugspurger replied:
Mmm probably an accident on my part. I'll push a fix (the rendered_html removal was intentional).

jreback commented:
@TomAugspurger how's this looking?

TomAugspurger commented:
I think it's good now. My last change removed some warnings thrown by IPython / pygments about
not being able to find the ipython lexer.

```diff
+              'IPython.sphinxext.ipython_console_highlighting',  # lowercase didn't work
```

We already had `ipython.sphinxext.ipython_console_highlighting` in there (lowercase ipython), but that apparently wasn't working. I have both there now, but we can probably remove the lowercased version?

mgeier commented:
Ideally, if IPython is installed properly, you wouldn't need to mention `ipython_console_highlighting` at all. Sadly, there seems to be a bug in Anaconda: https://github.com/ContinuumIO/anaconda-issues/issues/1430 (see also https://github.com/spatialaudio/nbsphinx/issues/24). I don't know if this is still broken or not.

And yes, the module name is (unfortunately) `IPython`, AFAIK there is no module called `ipython` (but it *should* be named like this).

jreback commented:
@TomAugspurger ok lmk. going to merge a bunch of things one after the other (because of annoying cache deleting once I merge the travis dedupe reversal).

jorisvandenbossche commented on doc/source/conf.py:
```diff
@@ -52,14 +52,16 @@
               'numpydoc', # used to parse numpy-style docstrings for autodoc
               'ipython_sphinxext.ipython_directive',
               'ipython_sphinxext.ipython_console_highlighting',
+              'IPython.sphinxext.ipython_console_highlighting',  # lowercase didn't work
```
Note there is a difference here in `_` vs `.`. The one already present with the `_` is our vendored version (and the directory is with lower case, so that works). What was the reason you needed this? 
(although we should maybe migrate to just use IPython itself instead of our own version, but should look again at why we exactly were doing this)

jreback commented:
can you rebase (happy to push this in as well if you are ready)

TomAugspurger commented:
Rebased.

> What was the reason you needed this?

Otherwise we got a bunch of warnings about the IPython lever not being
installed.
Ideally, we wouldn't need it but there's an issue with how anaconda
installs the notebook [apparently](
https://github.com/spatialaudio/nbsphinx/issues/24).

On Wed, Mar 29, 2017 at 2:38 PM, Jeff Reback <notifications@github.com>
wrote:

> can you rebase (happy to push this in as well if you are ready)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/15581#issuecomment-290201777>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABQHIj1JBV4z31ZJdGmrnmk1rHtrhTRdks5rqrMxgaJpZM4MTfzT>
> .
>

jreback commented on doc/source/contributing.rst:
```diff
@@ -359,15 +359,14 @@ have ``sphinx`` and ``ipython`` installed. `numpydoc
 requirements with::
 
       conda install -n pandas_dev sphinx ipython nbconvert nbformat
+      conda install -n pandas_dev -c conda-forge nbsphinx
```
we should just always add ``-c conda-forge`` (as defaults is always added), and things *just work*

jreback commented:
@TomAugspurger merge? (minor comment on the installation of libs).

jreback commented:
ready for merge?

TomAugspurger commented:
Going to wait a day or two to see if https://github.com/pandas-dev/pandas/pull/14757 is updated, since there will be a notebook conflict.

jreback commented:
sure

jreback commented:
nice @TomAugspurger couple of comments on the built docs: http://pandas-docs.github.io/pandas-docs-travis/style.html

- maybe need some ``:okwarning:`` to eliminate some of the displayed warnings (or slight code changes)
- can you align the numbers in the tables (maybe make this the first table)

jnothman commented:
The [documentation built with nbsphinx](http://pandas-docs.github.io/pandas-docs-travis/style.html) uses a different font to the rest of the docs, i.e. a serif font.

TomAugspurger commented:
Whoops, that's on me. Actually comes from my #15954 PR. I'll put up a fix. Thanks.
